bpf/bpf_netdev.c

```diff
@@ -97,6 +116,224 @@ static inline __u32 derive_sec_ctx(struct __sk_buff *skb, const union v6addr *no
+			ret = DROP_MISSED_TAIL_CALL;
+		}
+		goto drop_err;
+	}
+
+	if (!revalidate_data(skb, &data, &data_end, &ip6)) {
+		ret = DROP_INVALID;
+		goto drop_err;
+	}
+
+	fib_params.family = AF_INET6;
+	fib_params.ifindex = NATIVE_DEV_IFINDEX;
+
+	ipv6_addr_copy((union v6addr *) &fib_params.ipv6_src, (union v6addr *) &ip6->saddr);
+	ipv6_addr_copy((union v6addr *) &fib_params.ipv6_dst, (union v6addr *) &ip6->daddr);
+
+	ret = fib_lookup(skb, &fib_params, sizeof(fib_params),
+			 BPF_FIB_LOOKUP_DIRECT | BPF_FIB_LOOKUP_OUTPUT);
+	if (ret != 0) {
+		ret = DROP_NO_FIB;
+		goto drop_err;
+	}
+
+	if (eth_store_daddr(skb, fib_params.dmac, 0) < 0) {
+		ret = DROP_WRITE_ERROR;
+		goto drop_err;
+	}
+	if (eth_store_saddr(skb, fib_params.smac, 0) < 0) {
+		ret = DROP_WRITE_ERROR;
+		goto drop_err;
+	}
+
+	return redirect(fib_params.ifindex, 0);
+drop_err:
+	return send_drop_notify_error(skb, 0, ret, TC_ACT_SHOT,
+				      dir == NAT_DIR_INGRESS ?
+				      METRIC_INGRESS : METRIC_EGRESS);
+}
+
+/* See nodeport_lb4(). */
+static inline int nodeport_lb6(struct __sk_buff *skb, __u32 src_identity)
+{
+	int ret, l3_off = ETH_HLEN, l4_off, hdrlen;
+	struct ipv6_ct_tuple tuple = {};
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	struct csum_offset csum_off = {};
+	struct lb6_service_v2 *svc;
+	struct lb6_key_v2 key = {};
+	struct ct_state ct_state_new = {};
+	struct ct_state ct_state = {};
+	__u32 monitor = 0;
+	__u16 service_port;
+
+	if (!revalidate_data(skb, &data, &data_end, &ip6))
+		return DROP_INVALID;
+
+	tuple.nexthdr = ip6->nexthdr;
+	ipv6_addr_copy(&tuple.daddr, (union v6addr *) &ip6->daddr);
+	ipv6_addr_copy(&tuple.saddr, (union v6addr *) &ip6->saddr);
+
+	hdrlen = ipv6_hdrlen(skb, l3_off, &tuple.nexthdr);
+	if (hdrlen < 0)
+		return hdrlen;
+
+	l4_off = l3_off + hdrlen;
+
+	ret = lb6_extract_key_v2(skb, &tuple, l4_off, &key, &csum_off, CT_EGRESS);
+	if (IS_ERR(ret)) {
+		if (ret == DROP_UNKNOWN_L4)
+			return TC_ACT_OK;
+		else
+			return ret;
+	}
+
+	service_port = bpf_ntohs(key.dport);
+	if (service_port < NODEPORT_PORT_MIN ||
+	    service_port > NODEPORT_PORT_MAX) {
+		if (service_port >= NODEPORT_PORT_MIN_NAT &&
+		    service_port <= NODEPORT_PORT_MAX_NAT) {
+			skb->cb[CB_NAT] = NAT_DIR_INGRESS;
+			skb->cb[CB_SRC_IDENTITY] = src_identity;
+			ep_tail_call(skb, CILIUM_CALL_IPV6_NODEPORT_NAT);
+			return DROP_MISSED_TAIL_CALL;
+		}
+		return TC_ACT_OK;
+	}
+
+	ct_state_new.orig_dport = key.dport;
+
+	if ((svc = lb6_lookup_service_v2(skb, &key)) != NULL) {
+		ret = lb6_local(get_ct_map6(&tuple), skb, l3_off, l4_off,
+				&csum_off, &key, &tuple, svc, &ct_state_new);
+		if (IS_ERR(ret))
+			return ret;
+	} else {
+		return TC_ACT_OK;
+	}
+
+	ret = ct_lookup6(get_ct_map6(&tuple), &tuple, skb, l4_off, CT_EGRESS,
+			 &ct_state, &monitor);
+	if (ret < 0)
+		return ret;
+
+	switch (ret) {
+	case CT_NEW:
+		ct_state_new.src_sec_id = SECLABEL;
+		ct_state_new.node_port = 1;
+		ret = ct_create6(get_ct_map6(&tuple), &tuple, skb, CT_EGRESS,
```

Comment:
Still brainstorming on this one. On the one hand it makes sense to me to have these separate, at least for node-external traffic. Internal nodeport traffic is handled via cgroups hooks, so it looks like regular connections to the data path, but it's also not too relevant in this context given other EP on the same node could spam the CT as well. Problem with external nodeport traffic where backend EP is on the same node is that the two CT map lookups will cause verifier complexity explosion in bpf_lxc and also sit in critical fast-path. Somehow it would be nice to have the same single LRU map and have different eviction policies, e.g. all internal traffic would be in one zone and all external traffic would be marked with a different one, so updates with a zone flag only ever evict entries in their own zone.